src/SAWScript/MGU.hs

```diff
@@ -67,44 +67,118 @@ listSubst = Subst . M.fromList
 
 -- Most General Unifier {{{
 
-failMGU :: String -> Either String a
-failMGU = Left
-
-assert :: Bool -> String -> Either String ()
-assert b msg = unless b $ failMGU msg
-
-mgu :: LName -> Type -> Type -> Either String Subst
-mgu m (LType _ t) t2 = mgu m t t2
-mgu m t1 (LType _ t) = mgu m t1 t
-mgu m (TyUnifyVar i) t2 = bindVar m i t2
-mgu m t1 (TyUnifyVar i) = bindVar m i t1
-mgu m r1@(TyRecord ts1) r2@(TyRecord ts2) = do
-  assert (M.keys ts1 == M.keys ts2) $
-    "Record field type mismatch. Expected " ++ pShow r1 ++ " but got " ++ pShow r2 ++ " at " ++ show m
-  mgus m (M.elems ts1) (M.elems ts2)
-mgu m (TyCon tc1 ts1) (TyCon tc2 ts2)
-  | tc1 == tc2 = (mgus m ts1 ts2)
+{-
+ - If we fail here, the caller attaches the message we return to one
+ - of its own that gives the position of the best enclosing expression
+ - position available. It's important to make it clear that this is
+ - all one message. Since what we send back is printed on its own
+ - lines, indent it four spaces (the caller's message is already
+ - indented two) to make it more clearly part of its preceding text.
+ -
+ - We're going to print the expected and found types on their own
+ - lines because they can be large. (If so they can still be fairly
+ - illegible, but at least the user doesn't have to hunt for "found"
+ - in the middle of a multi-line print.) Also pad the prefix of the
+ - prints so that the types line up; this is helpful for longer types
+ - that still fit on one output line.
+ -
+ - We have recursed into the types and so if we get a mismatch we
+ - should print those types (this is important if they are e.g.
+ - elements in a large system of nested records and tuples) but also
+ - print the enclosing types on the way out. This can become rather
+ - verbose but makes it a lot easier to deal with errors deep within
+ - complex structures.
+ -
+ - The failure result is a start line, then all the expected/found
+ - lines we've accumulated, and finally, the in the middle, and the
+ - last seen expected/found lines arising from a function type, which
+ - we don't commit to using until we see a non-function type or get to
+ - the end. The idea is that if we pass through a chain of function
+ - types, each having the curried next argument to the previous, we
+ - only print the last and largest one.
+ -
+ - Note that although we append to the end of the expected/found list,
+ - we don't stick the start line in that list, because I keep going
+ - back and forth on whether the larger types should be printed first
+ - (prepending in failMGUadd) or last (appending). If we commit to
+ - appending we don't need to keep the start line separate.
+ -}
+
+type FailMGU = (String, [String], Maybe [String])
+
+-- common code for printing expected/found types
+showtypes :: Type -> Type -> [String]
+showtypes ty1 ty2 =
```

Comment:
I'd prefer to stick to the camelCase conventions used elsewhere in SAW, so let's name this `showTypes` instead.

Reply:
ok... fixed in 1b48e419c7d418b77a8daed0a62095e7d0d4e4f3